src/plugins/intel_gpu/src/plugin/remote_tensor.cpp

```diff
@@ -14,6 +14,117 @@
 namespace ov {
 namespace intel_gpu {
 
+namespace {
+static void calculate_strides(ov::Strides& strides, const ov::Shape& shape, const ov::element::Type& element_type) {
+    if (element_type.bitwidth() < 8)
+        return;
+    strides.clear();
+    if (!shape.empty()) {
+        strides.resize(shape.size());
+        strides.back() = shape.back() == 0 ? 0 : element_type.size();
+        std::copy(shape.rbegin(), shape.rend() - 1, strides.rbegin() + 1);
+        std::partial_sum(strides.rbegin(), strides.rend(), strides.rbegin(), std::multiplies<size_t>());
+    }
+}
+
+struct mem_wrapper {
+    mem_wrapper(cldnn::stream& stream, cldnn::memory_ptr mem_ptr, void* data_ptr)
+        : m_stream(stream)
+        , m_mem_ptr(mem_ptr)
+        , m_data_ptr(data_ptr) {
+            OPENVINO_ASSERT((m_data_ptr != nullptr) != (m_mem_ptr != nullptr), "[GPU] Invalid memory configuration");
+        }
+
+    void copy_to(mem_wrapper& dst, size_t src_offset, size_t dst_offset, size_t size) const {
+        const bool is_blocking = true;
+        if (m_data_ptr != nullptr) {
+            OPENVINO_ASSERT(dst.m_mem_ptr, "[GPU] Unexpected host to host copy call for Remote Tensors");
+
+            // Device <== Host
+            dst.m_mem_ptr->copy_from(m_stream, m_data_ptr, src_offset, dst_offset, size, is_blocking);
+        } else {
+            if (dst.m_data_ptr != nullptr) {
+                // Device ==> Host
+                m_mem_ptr->copy_to(m_stream, dst.m_data_ptr, src_offset, dst_offset, size, is_blocking);
+            } else {
+                // Device ==> Device
+                m_mem_ptr->copy_to(m_stream, *dst.m_mem_ptr, src_offset, dst_offset, size, is_blocking);
+            }
+        }
+    }
+
+private:
+    cldnn::stream& m_stream;
+    cldnn::memory_ptr m_mem_ptr = nullptr;
+    void* m_data_ptr = nullptr;
+};
+
+static void copy_roi_recursively(const mem_wrapper& src_mem,
+                     mem_wrapper& dst_mem,
+                     const size_t axis,
+                     const size_t src_offset,
+                     const size_t dst_offset,
+                     const ov::Shape& roi_shape,
+                     const ov::Strides& src_strides,
+                     const ov::Strides& dst_strides,
+                     const ov::Strides& roi_strides) {
+    if (axis == roi_shape.size() - 1) {
+        // Copy the innermost dimension
+        const auto size = roi_strides[axis] * roi_shape[axis];
+        src_mem.copy_to(dst_mem, src_offset, dst_offset, size);
+    } else {
+        // Check if the current dimension and all inner dimensions can be copied as a single chunk
+        bool can_copy_as_chunk = true;
+        for (size_t i = axis; i < roi_shape.size(); i++) {
+            if (src_strides[i] != roi_strides[i] || dst_strides[i] != roi_strides[i]) {
```

Comment:
Maybe we can simplify that a little bit?
I was thinking about something like this:
1. calculate squeezed shape. i.e. try to merge adjacent dimensions if there's no padding for them (might be even precalculated on tensor construction to avoid overhead on each copy_* call
2. just loop over squeezed shape and call copy for innermost dimension

Reply:
Can we stick with the current approach? I tried your suggestion, but it seems like it requires extra stuff like squeezed_src/dst/roi shapes and updated squeezed strides for all of them to navigate the tensors properly. The copy logic is clearer, but the preparation gets more complicated, so it doesn't seem much simpler overall...

Reply:
Ok, we can keep it as is. But I think it would be good to measure overhead of this method for the copy of small data chunks